### PR TITLE
Initial bluez backend

### PR DESCRIPTION
**Pull request by abrauchli**
Here's a new backend implementation based on bluez's official dbus API to complement the gatttool backend. pydbus is used as dbus library. Comments welcome.

### CODE REVIEW

**Comment by danjjl** [comment]
It would be really nice to have a dbus backend!

You could add `python3-gi` to the `travis.yml`:

```
before_install:
  - sudo apt-get update
  - sudo apt-get install -qq python3-gi
```

**Comment by abrauchli** [comment]
Thanks for your suggestion @danjjl - that probably needs a python with _system_site_package and the current precise (12.04) does't have it for py3.5

I took an initial stab at the unit tests. Although they're not complete yet, I'm not sure when I'll find time again so I rebased the changeset where it's at. Feel free to continue.

File: pygatt/backends/bluez/bluez.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+                                            pattern)
+
+    def find_adapter_in_objects(self, objects, pattern=None):
+        for path, ifaces in objects.iteritems():
+            adapter = ifaces.get(DBusHelper.ADAPTER_INTERFACE)
+            if adapter is None:
+                continue
+
+            try:
+                if (not pattern and adapter.Powered) or \
```
Comment (yawor) [inline pygatt/backends/bluez/bluez.py]: I've found that (at least under Python 2.7), `adapter.Powered` is not evaluated to `True` and no devices are returned, even when the device is powered. After changing to `adapter['Powered']` the first powered adapter is returned properly.

File: pygatt/backends/bluez/device.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        bus_obj = self._dbus.get(self._dbus.SERVICE_NAME, self._dbus_path,
+                                 timeout=timeout)
+        bus_obj.Disconnect()
+        self._connected = False
+        log.info("Disconnected from %s", self.address)
+
+    def discover_characteristics(self, bledevice,
+                                 timeout=DEFAULT_CONNECT_TIMEOUT_S):
+        dbus_obj = self._dbus.object_by_path(self._dbus_path,
+                                            interface=self.DEVICE_INTERFACE)
```
Comment (yawor) [inline pygatt/backends/bluez/device.py]: There's no `self.DEVICE_INTERFACE`. It should be `self._dbus.DEVICE_INTERFACE`.

File: pygatt/backends/bluez/device.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            dbus_obj = self._dbus.object_by_path(path,
+                    interface=self._dbus.GATT_CHAR_INTERFACE)
+            v = dbus_obj.ReadValue({})
+            return bytearray(v)
+        raise Exception("UUID {} not found".format(uuid))
+
+    @connection_required
+    def char_write(self, uuid, value, wait_for_response=False):
+        objs = self._dbus.objects_by_property({'UUID': uuid},
+                                                      base_path='/')
```
Comment (yawor) [inline pygatt/backends/bluez/device.py]: The `objects_by_property` method is missing an interface argument here: `interface=self._dbus.GATT_CHAR_INTERFACE`.

**Comment by peplin** [comment]
I apologize for letting this sit for too long. If you still are interested, there's another implementation in #239, and a more actively maintained library here: https://github.com/hbldh/bleak

**Comment by abrauchli** [comment]
No worries, thanks for your code and the recommendation: Bleak is great, I've been using that for a while now.

**Pull request closed without merging**